File: packages/server/src/domains/audit/service.ts
import type { AuditAction, AuditEntityType, AuditEntry, SessionUser } from '@flick/shared';
import * as repo from './repo.js';

export async function log(
  projectId: string,
  actor: SessionUser,
  action: AuditAction,
  entityType: AuditEntityType,
  entityId: string,
  entityName: string,
  before: Record<string, unknown> | null,
  after: Record<string, unknown> | null,
): Promise<AuditEntry> {
  return repo.create({
    project_id: projectId,
    actor_id: actor.id,
    actor_email: actor.email,
    action,
    entity_type: entityType,
    entity_id: entityId,
    entity_name: entityName,
    before,
    after,
  });
}

export async function getAuditLog(
  projectId: string,
  opts?: {
    entity_type?: string;
    entity_id?: string;
    actor_id?: string;
    action?: string;
    cursor?: string;
    limit?: number;
  },
) {
  return repo.findByProject(projectId, opts);
}

File: packages/server-rust/src/domains/audit/service.rs
use serde_json::Value;
use sqlx::PgPool;
use uuid::Uuid;

use crate::auth::AuthUser;
use crate::error::AppError;

use super::repo::{self, AuditEntryRow};

pub async fn log(
    db: &PgPool,
    project_id: Uuid,
    actor: &AuthUser,
    action: &str,
    entity_type: &str,
    entity_id: &str,
    entity_name: &str,
    before_state: Option<Value>,
    after_state: Option<Value>,
) -> Result<AuditEntryRow, AppError> {
    let actor_id: Uuid = actor.id.parse().map_err(|_| AppError::validation("Invalid actor ID"))?;
    repo::create(
        db,
        project_id,
        actor_id,
        &actor.email,
        action,
        entity_type,
        entity_id,
        entity_name,
        before_state,
        after_state,
    )
    .await
    .map_err(AppError::from)
}

pub async fn get_audit_log(
    db: &PgPool,
    project_id: Uuid,
    entity_type: Option<&str>,
    entity_id: Option<&str>,
    actor_id: Option<Uuid>,
    action: Option<&str>,
    cursor: Option<&str>,
    limit: Option<i64>,
) -> Result<(Vec<AuditEntryRow>, Option<String>, bool), AppError> {
    repo::find_by_project(
        db,
        project_id,
        entity_type,
        entity_id,
        actor_id,
        action,
        cursor,
        limit,
    )
    .await
    .map_err(AppError::from)
}
